# Determining Car Specifications and Prices

## Project Description
Developing a linear regression model to predict car prices by analyzing the correlation between physical specifications and engine performance. This project covers the entire data science pipeline, from data cleaning and outlier handling to model evaluation using the R² score and Root Mean Squared Error (RMSE) metrics to ensure reliable price estimates.

---

Mengembangkan model regresi linear untuk memprediksi harga mobil dengan menganalisis korelasi antara spesifikasi fisik dan performa mesin. Proyek ini mencakup seluruh pipeline data science, mulai dari pembersihan data, penanganan outlier, hingga evaluasi model menggunakan metrik R2-Score dan Root Mean Squared Error (RMSE) untuk memastikan estimasi harga yang kredibel.

## Dataset
- **Source**: [Kaggle - Car Sales Dataset](https://www.kaggle.com/datasets/gagandeep16/car-sales)
- **Size**: 157 rows, 16 columns
- **Target**: `Price_in_thousands` column (Continuous numerical value in $1,000s)


---

- **Sumber**: [Kaggle - Car Sales Dataset](https://www.kaggle.com/datasets/gagandeep16/car-sales)
- **Ukuran**: 157 baris, 16 kolom
- **Target**: `Price_in_thousands` kolom (Nilai numerik kontinu di $1,000)

## Project Workflow
1. **Data Understanding** — Exploring the structure and characteristics of the dataset
2. **Data Preparation** — Handling missing values.
3. **Exploratory Data Analysis (EDA)** — Visualizing patterns and extracting insights
4. **Modelling** — Building a prediction model using LinearRegression
5. **Evaluation** — Measuring model performance using R2-Score, Root Mean Squared Error(RMSE)


---

1. **Pemahaman Data** — Menjelajahi struktur dan karakteristik kumpulan data
2. **Persiapan Data** — Menangani nilai yang hilang.
3. **Analisis Data Eksploratori (EDA)** — Memvisualisasikan pola dan mengekstrak wawasan
4. **Pemodelan** — Membangun model prediksi menggunakan LinearRegression
5. **Evaluasi** — Mengukur kinerja model menggunakan R2-Score, Root Mean Squared Error (RMSE)

## Tools & Libraries
- Python, Pandas, NumPy
- Plotly Express (data visualization)
- Scikit-learn (machine learning)


---

- Python, Pandas, NumPy
- Plotly Express (visualisasi data)
- Scikit-learn (pembelajaran mesin)

## Data Understanding

In [1]:
# Import library
import pandas as pd
import numpy as np
import plotly.express as px

In [2]:
# Load Data
df = pd.read_csv("Car_sales.csv")
df

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
0,Acura,Integra,16.919,16.360,Passenger,21.50,1.8,140.0,101.2,67.3,172.4,2.639,13.2,28.0,2/2/2012,58.280150
1,Acura,TL,39.384,19.875,Passenger,28.40,3.2,225.0,108.1,70.3,192.9,3.517,17.2,25.0,6/3/2011,91.370778
2,Acura,CL,14.114,18.225,Passenger,NaN,3.2,225.0,106.9,70.6,192.0,3.470,17.2,26.0,1/4/2012,NaN
3,Acura,RL,8.588,29.725,Passenger,42.00,3.5,210.0,114.6,71.4,196.6,3.850,18.0,22.0,3/10/2011,91.389779
4,Audi,A4,20.397,22.255,Passenger,23.99,1.8,150.0,102.6,68.2,178.0,2.998,16.4,27.0,10/8/2011,62.777639
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
152,Volvo,V40,3.545,NaN,Passenger,24.40,1.9,160.0,100.5,67.6,176.6,3.042,15.8,25.0,9/21/2011,66.498812
153,Volvo,S70,15.245,NaN,Passenger,27.50,2.4,168.0,104.9,69.3,185.9,3.208,17.9,25.0,11/24/2012,70.654495
154,Volvo,V70,17.531,NaN,Passenger,28.80,2.4,168.0,104.9,69.3,186.2,3.259,17.9,25.0,6/25/2011,71.155978
155,Volvo,C70,3.493,NaN,Passenger,45.50,2.3,236.0,104.9,71.5,185.7,3.601,18.5,23.0,4/26/2011,101.623357


In [3]:
# detail information from dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Manufacturer         157 non-null    object 
 1   Model                157 non-null    object 
 2   Sales_in_thousands   157 non-null    float64
 3   __year_resale_value  121 non-null    float64
 4   Vehicle_type         157 non-null    object 
 5   Price_in_thousands   155 non-null    float64
 6   Engine_size          156 non-null    float64
 7   Horsepower           156 non-null    float64
 8   Wheelbase            156 non-null    float64
 9   Width                156 non-null    float64
 10  Length               156 non-null    float64
 11  Curb_weight          155 non-null    float64
 12  Fuel_capacity        156 non-null    float64
 13  Fuel_efficiency      154 non-null    float64
 14  Latest_Launch        157 non-null    object 
 15  Power_perf_factor    155 non-null    flo

In [4]:
# Data type
df.dtypes

,0
Manufacturer,object
Model,object
Sales_in_thousands,float64
__year_resale_value,float64
Vehicle_type,object
Price_in_thousands,float64
Engine_size,float64
Horsepower,float64
Wheelbase,float64
Width,float64


In [5]:
# view the number of unique values
df.nunique()

,0
Manufacturer,30
Model,156
Sales_in_thousands,157
__year_resale_value,117
Vehicle_type,2
Price_in_thousands,152
Engine_size,31
Horsepower,66
Wheelbase,88
Width,78


In [6]:
# view the number of missing values
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,2
Engine_size,1
Horsepower,1
Wheelbase,1
Width,1


## Data Preparation

In [7]:
# viewing data with null values in the Engine_size feature
df[df['Engine_size'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
33,Chrysler,Town & Country,53.48,19.54,Car,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,7/13/2011,NaN


In [8]:
# deleting records with a null value in the Engine_size column
df = df.drop(df[df['Engine_size'].isna() == True].index)
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,1
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


In [9]:
# viewing data with null values in the Price_in_thousands feature
df[df['Price_in_thousands'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
2,Acura,CL,14.114,18.225,Passenger,NaN,3.2,225.0,106.9,70.6,192.0,3.47,17.2,26.0,1/4/2012,NaN


In [10]:
# imputation of values using the mean for the Price_in_thousands feature
df['Price_in_thousands'] = df['Price_in_thousands'].fillna(df['Price_in_thousands'].mean())
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,0
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


In [11]:
# viewing data with null values in the Curb_weight feature
df[df['Curb_weight'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
15,Cadillac,Seville,15.943,27.1,Passenger,44.475,4.6,275.0,112.2,75.0,201.0,NaN,18.5,22.0,4/29/2011,115.621358


In [12]:
# imputation of values using the mean for the Curb_weight feature
df['Curb_weight'] = df['Curb_weight'].fillna(df['Curb_weight'].mean())
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,0
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


In [13]:
# viewing data with null values in the Fuel_efficiency feature
df[df['Fuel_efficiency'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
38,Dodge,Intrepid,88.028,12.275,Passenger,22.505,2.7,202.0,113.0,74.7,203.7,3.489,17.0,NaN,6/2/2012,80.831470
109,Oldsmobile,Intrigue,38.554,NaN,Passenger,24.150,3.5,215.0,109.0,73.6,195.9,3.455,18.0,NaN,4/1/2011,86.272523


In [14]:
# imputation of values using the mean for the Fuel_efficiency feature
df['Fuel_efficiency'] = df['Fuel_efficiency'].fillna(df['Fuel_efficiency'].mean())
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,0
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


In [15]:
# viewing data with null values in the Curb_weight feature
df[df['Power_perf_factor'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
2,Acura,CL,14.114,18.225,Passenger,27.390755,3.2,225.0,106.9,70.6,192.0,3.47,17.2,26.0,1/4/2012,NaN


In [16]:
# imputation of values using the mean for the Power_perf_factor feature
df['Power_perf_factor'] = df['Power_perf_factor'].fillna(df['Power_perf_factor'].mean())
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,36
Vehicle_type,0
Price_in_thousands,0
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


In [17]:
# viewing data with null values in the __year_resale_value feature
df[df['__year_resale_value'].isna() == True]

,Manufacturer,Model,Sales_in_thousands,__year_resale_value,Vehicle_type,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Latest_Launch,Power_perf_factor
7,BMW,323i,19.747,NaN,Passenger,26.990,2.5,170.0,107.3,68.4,176.0,3.179,16.6,26.000000,6/28/2011,71.191207
18,Cadillac,Escalade,14.785,NaN,Car,46.225,5.7,255.0,117.5,77.0,201.2,5.572,30.0,15.000000,4/17/2012,109.509117
27,Chevrolet,Impala,107.995,NaN,Passenger,18.890,3.4,180.0,110.5,73.0,200.0,3.389,17.0,27.000000,6/18/2011,71.838039
34,Chrysler,300M,30.696,NaN,Passenger,29.185,3.5,253.0,113.0,74.4,197.8,3.567,17.0,23.000000,2/10/2012,101.655244
44,Dodge,Durango,101.323,NaN,Car,26.310,5.2,230.0,115.7,71.7,193.5,4.394,25.0,17.000000,6/27/2012,92.854125
50,Ford,Focus,175.670,NaN,Passenger,12.315,2.0,107.0,103.0,66.9,174.8,2.564,13.2,30.000000,7/22/2012,43.117132
66,Jaguar,S-Type,15.467,NaN,Passenger,42.800,3.0,240.0,114.5,71.6,191.3,3.650,18.4,21.000000,11/3/2012,102.178985
72,Lexus,GS400,3.334,NaN,Passenger,46.305,4.0,300.0,110.2,70.9,189.2,3.693,19.8,21.000000,11/28/2012,125.013357
74,Lexus,LX470,9.126,NaN,Car,60.105,4.7,230.0,112.2,76.4,192.5,5.401,25.4,15.000000,10/30/2012,105.760458
75,Lexus,RX300,51.238,NaN,Car,34.605,3.0,220.0,103.0,71.5,180.1,3.900,17.2,21.000000,1/4/2012,91.943802


In [18]:
# the difference between the new price and the used price
df[['Price_in_thousands', '__year_resale_value']]

,Price_in_thousands,__year_resale_value
0,21.500000,16.360
1,28.400000,19.875
2,27.390755,18.225
3,42.000000,29.725
4,23.990000,22.255
...,...,...
152,24.400000,NaN
153,27.500000,NaN
154,28.800000,NaN
155,45.500000,NaN


In [19]:
# create a variable that stores the average difference between the prices of new and used cars
reduction_average = pd.Series(df['Price_in_thousands'] - df['__year_resale_value']).mean()
reduction_average

np.float64(8.045922956989248)

In [20]:
# imputing values for the __year_resale_value feature
df['__year_resale_value'] = df['__year_resale_value'].fillna(df['Price_in_thousands'].sub(reduction_average))
df.isna().sum()

,0
Manufacturer,0
Model,0
Sales_in_thousands,0
__year_resale_value,0
Vehicle_type,0
Price_in_thousands,0
Engine_size,0
Horsepower,0
Wheelbase,0
Width,0


## Exploratory Data Analysis (EDA)

#### View the best-selling models

In [21]:
# view the best-selling models
df_model = df[['Model', 'Manufacturer', 'Sales_in_thousands']].groupby(['Model', 'Manufacturer']).sum().reset_index()
df_model = df_model.sort_values(['Sales_in_thousands'], ascending = False)
df_model.head(10)

,Model,Manufacturer,Sales_in_thousands
63,F-Series,Ford,540.561
62,Explorer,Ford,276.747
31,Camry,Toyota,247.994
146,Taurus,Ford,245.815
12,Accord,Honda,230.902
118,Ram Pickup,Dodge,227.061
121,Ranger,Ford,220.650
41,Civic,Honda,199.685
32,Caravan,Dodge,181.749
65,Focus,Ford,175.670


In [22]:
fig = px.bar(df_model.head(10), x='Model', y='Sales_in_thousands', color='Model', text_auto=True, title='10 Model dengan Jumlah Penjualan Terbaik')
fig.update_traces(textfont_size=10, textposition='inside')
fig.update_layout(width=1000, height=800)
fig.show()

> Dapat kita lihat, di atas merupakan sepuluh model mobil yang memiliki penjualan terbaik di pasaran. F-Series memimpin dengan lebih dari 500 ribu penjualan, disusul oleh model Explorer dan Camry. Dilihat dari manufaktur, Ford mendominasi daftar sepuluh penjualan terbaik dengan 5 model mobilnya. Nantinya kesepuluh model ini dapat menjadi acuan kita, mobil seperti apa yang terbukti laris di pasaran





#### Create a dataframe to analyze the model's features

In [23]:
col_selected =['Model', 'Vehicle_type', '__year_resale_value', 'Price_in_thousands', 'Engine_size', 'Horsepower', 'Wheelbase', 'Width', 'Length', 'Curb_weight', 'Fuel_capacity', 'Fuel_efficiency', 'Power_perf_factor']

In [24]:
df_selected = df[col_selected].groupby(['Model', 'Vehicle_type']).mean().reset_index()
df_selected

,Model,Vehicle_type,__year_resale_value,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Power_perf_factor
0,3-Sep,Passenger,18.054077,26.100,2.0,185.0,102.6,67.4,182.2,2.990,16.9,23.0,76.023048
1,3000GT,Passenger,20.940000,25.450,3.0,161.0,97.2,72.4,180.3,3.131,19.8,21.0,67.544155
2,300M,Passenger,21.139077,29.185,3.5,253.0,113.0,74.4,197.8,3.567,17.0,23.0,101.655244
3,323i,Passenger,18.944077,26.990,2.5,170.0,107.3,68.4,176.0,3.179,16.6,26.0,71.191207
4,328i,Passenger,28.675000,33.400,2.8,193.0,107.3,68.5,176.0,3.197,16.6,24.0,81.877069
...,...,...,...,...,...,...,...,...,...,...,...,...,...
150,Viper,Passenger,58.470000,69.725,8.0,450.0,96.2,75.7,176.7,3.375,19.0,16.0,188.144323
151,Voyager,Car,12.025000,18.850,2.4,150.0,113.3,76.8,186.3,3.528,20.0,24.0,60.951185
152,Windstar,Car,13.175000,21.410,3.0,150.0,120.7,76.6,200.9,3.761,26.0,21.0,62.095048
153,Wrangler,Car,13.475000,14.460,2.5,120.0,93.4,66.7,152.0,3.045,19.0,17.0,48.672898


In [25]:
df_to_analyze = df_selected.merge(df_model, on='Model')
df_to_analyze.sort_values(['Sales_in_thousands'], ascending=False, inplace=True)
df_to_analyze.head(10)

,Model,Vehicle_type,__year_resale_value,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Power_perf_factor,Manufacturer,Sales_in_thousands
63,F-Series,Car,15.075000,26.935,4.6,220.0,138.5,79.1,224.5,4.241,25.1,18.0,89.401935,Ford,540.561
62,Explorer,Car,16.640000,31.930,4.0,210.0,111.6,70.2,190.7,3.876,21.0,19.0,87.635496,Ford,276.747
31,Camry,Passenger,13.245000,17.518,2.2,133.0,105.2,70.1,188.5,2.998,18.5,27.0,54.372420,Toyota,247.994
146,Taurus,Passenger,10.055000,17.885,3.0,155.0,108.5,73.0,197.6,3.368,16.0,24.0,62.503740,Ford,245.815
12,Accord,Passenger,13.210000,15.350,2.3,135.0,106.9,70.3,188.8,2.932,17.1,27.0,54.269548,Honda,230.902
118,Ram Pickup,Car,15.060000,19.460,5.2,230.0,138.7,79.3,224.2,4.470,26.0,17.0,90.211700,Dodge,227.061
121,Ranger,Car,7.850000,12.050,2.5,119.0,117.5,69.4,200.7,3.086,20.0,23.0,47.389531,Ford,220.650
41,Civic,Passenger,9.850000,12.885,1.6,106.0,103.2,67.1,175.1,2.339,11.9,32.0,42.879097,Honda,199.685
32,Caravan,Car,12.025000,19.565,2.4,150.0,113.3,76.8,186.3,3.533,20.0,24.0,61.227000,Dodge,181.749
65,Focus,Passenger,4.269077,12.315,2.0,107.0,103.0,66.9,174.8,2.564,13.2,30.0,43.117132,Ford,175.670


#### Prices of the best-selling cars

In [26]:
fig = px.bar(df_to_analyze.head(10), x='Price_in_thousands', y='Model', color='Model', orientation='h', text_auto=True, title='Harga dari 10 model terbaik')
fig.update_traces(textfont_size=10, textposition='outside')
fig.show()



> Grafik diatas menampilkan 10 model mobil dengan harga terbaik. Disini Explorer memiliki harga terbaik dengan harga 31.930, disusul dengan F-Series di posisi kedua dengan harga 26.935, dan disisi lain terdapat Ranger dengan pilihan yang paling terjangkau dengan harga terendah yaitu 12.050. Ford menunjukkan dominasi luar biasa di antara model-model terbaik ini dengan menempatkan lima model mobilnya (F-Series, Explorer, Taurus, Ranger, dan Focus) ke dalam daftar sepuluh besar. Nantinya, data harga ini dapat dikolaborasikan dengan metrik kinerja lainnya, seperti volume penjualan atau kepuasan pelanggan, untuk memberikan analisis yang lebih komprehensif tentang mengapa model-model khusus ini dianggap sebagai 'terbaik' di pasar.



#### The horsepower rating of the best-selling model

In [27]:
fig = px.bar(df_to_analyze.head(10), x='Horsepower', y='Model', color='Model', orientation='h', text_auto=True, title='Angka Horsepower dari 10 model dengan penjualan terbaik')
fig.update_traces(textfont_size=10, textposition='outside')
fig.show()

#### Determining the specifications of the car to be produced

In [28]:
df_to_analyze.head(10)

,Model,Vehicle_type,__year_resale_value,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Power_perf_factor,Manufacturer,Sales_in_thousands
63,F-Series,Car,15.075000,26.935,4.6,220.0,138.5,79.1,224.5,4.241,25.1,18.0,89.401935,Ford,540.561
62,Explorer,Car,16.640000,31.930,4.0,210.0,111.6,70.2,190.7,3.876,21.0,19.0,87.635496,Ford,276.747
31,Camry,Passenger,13.245000,17.518,2.2,133.0,105.2,70.1,188.5,2.998,18.5,27.0,54.372420,Toyota,247.994
146,Taurus,Passenger,10.055000,17.885,3.0,155.0,108.5,73.0,197.6,3.368,16.0,24.0,62.503740,Ford,245.815
12,Accord,Passenger,13.210000,15.350,2.3,135.0,106.9,70.3,188.8,2.932,17.1,27.0,54.269548,Honda,230.902
118,Ram Pickup,Car,15.060000,19.460,5.2,230.0,138.7,79.3,224.2,4.470,26.0,17.0,90.211700,Dodge,227.061
121,Ranger,Car,7.850000,12.050,2.5,119.0,117.5,69.4,200.7,3.086,20.0,23.0,47.389531,Ford,220.650
41,Civic,Passenger,9.850000,12.885,1.6,106.0,103.2,67.1,175.1,2.339,11.9,32.0,42.879097,Honda,199.685
32,Caravan,Car,12.025000,19.565,2.4,150.0,113.3,76.8,186.3,3.533,20.0,24.0,61.227000,Dodge,181.749
65,Focus,Passenger,4.269077,12.315,2.0,107.0,103.0,66.9,174.8,2.564,13.2,30.0,43.117132,Ford,175.670


In [29]:
df_prediction = df_to_analyze.drop(['Model', 'Vehicle_type', '__year_resale_value', 'Price_in_thousands', 'Manufacturer', 'Sales_in_thousands'], axis=1)
df_prediction.mean().round()

,0
Engine_size,3.0
Horsepower,186.0
Wheelbase,107.0
Width,71.0
Length,187.0
Curb_weight,3.0
Fuel_capacity,18.0
Fuel_efficiency,24.0
Power_perf_factor,77.0


## Modelling

#### Import Library

In [30]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error, r2_score

#### Feature Selection

In [31]:
pred_data = df_to_analyze.drop(['Model', 'Manufacturer', '__year_resale_value', 'Sales_in_thousands', 'Vehicle_type'], axis=1)
pred_data

,Price_in_thousands,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Power_perf_factor
63,26.935,4.6,220.0,138.5,79.1,224.5,4.241,25.1,18.0,89.401935
62,31.930,4.0,210.0,111.6,70.2,190.7,3.876,21.0,19.0,87.635496
31,17.518,2.2,133.0,105.2,70.1,188.5,2.998,18.5,27.0,54.372420
146,17.885,3.0,155.0,108.5,73.0,197.6,3.368,16.0,24.0,62.503740
12,15.350,2.3,135.0,106.9,70.3,188.8,2.932,17.1,27.0,54.269548
...,...,...,...,...,...,...,...,...,...,...
34,71.020,3.4,300.0,92.6,69.5,174.5,3.032,17.0,21.0,134.390975
49,18.145,3.1,150.0,107.0,69.4,192.0,3.102,15.2,25.0,60.861612
26,85.500,5.0,302.0,113.6,73.1,196.6,4.115,23.2,20.0,141.100985
151,69.725,8.0,450.0,96.2,75.7,176.7,3.375,19.0,16.0,188.144323


In [32]:
x = pred_data.drop('Price_in_thousands', axis=1)
y = pred_data['Price_in_thousands']

In [33]:
x.shape

(156, 9)

#### Train Test Split

In [34]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

#### Create a regression model

In [35]:
model = LinearRegression()
model.fit(x_train, y_train)

LinearRegression()

#### Conduct a performance evaluation

In [36]:
y_pred = model.predict(x_test)
y_pred

array([22.51570812, 21.50763923, 25.68429089, 31.9670488 , 37.4895899 ,
       10.40542596, 11.61339168, 40.15858797, 47.01214156, 16.55553592,
       20.19308323, 18.96082481, 19.8607866 , 50.09218599, 23.96486346,
       22.36244924, 29.02608831, 33.03537774, 17.23293844, 37.46063521,
       26.39221772, 12.72358899, 12.08795701, 11.92785315, 18.33207148,
       28.67175964, 17.95854315, 24.29503006, 22.04825048, 31.13538758,
       21.35889437, 40.27446622])

In [37]:
rmse = root_mean_squared_error(y_test, y_pred)
r2score = r2_score(y_test, y_pred)
print('RMSE = ', rmse)
print('R2 Score = ', r2score)

RMSE =  1.7114743355200899
R2 Score =  0.971077700265979


In [38]:
results_df = pd.DataFrame({'Actual Sales': y_test, 'Predicted Sales': y_pred})
results_df

,Actual Sales,Predicted Sales
150,22.510,22.515708
66,20.095,21.507639
114,26.399,25.684291
2,29.185,31.967049
15,36.229,37.489590
129,10.685,10.405426
94,9.235,11.613392
116,42.000,40.158588
103,42.660,47.012142
70,17.500,16.555536


In [39]:
fig = px.scatter(results_df, x='Predicted Sales', y='Actual Sales', trendline='ols')
fig.update_layout(
    title='Actual vs Predicted Sales',
    xaxis_title='Predicted Sales',
    yaxis_title='Actual Sales'
)
fig.show()

#### Make price predictions for cars with specified specifications

In [41]:
data = {
    'Engine_size': [3.0],
    'Horsepower': [186.0],
    'Wheelbase': [107.0],
    'Width': [71.0],
    'Length': [187.0],
    'Curb_weight': [3.0],
    'Fuel_capacity': [18.0],
    'Fuel_efficiency': [24.0],
    'Power_perf_factor': [77.0]
}

df_pred_price = pd.DataFrame(data)
df_pred_price

,Engine_size,Horsepower,Wheelbase,Width,Length,Curb_weight,Fuel_capacity,Fuel_efficiency,Power_perf_factor
0,3.0,186.0,107.0,71.0,187.0,3.0,18.0,24.0,77.0


In [42]:
price_pred = model.predict(df_pred_price)
print('Harga mobil : {:.2f}'.format(price_pred[0]))

Harga mobil : 26.40
